### PR TITLE
deps(yamux): update yamux to `v0.12`

### PR DESCRIPTION
**Pull request by thomaseizinger**
## Notes

~As discussed in https://github.com/libp2p/rust-yamux/pull/142#issuecomment-1318082284, I would suggest to release this as an alpha and then [ask our users](https://github.com/libp2p/rust-libp2p/issues/3202) to depend on that alpha directly instead of using `libp2p::yamux` so we can get some feedback on how this implementation performs in the wild.~

~I think we should probably hold off on merging this but keep this PR to discuss any feedback that we might get. Perhaps the changelog entry should be merged to ensure that one is correct.~

~Thoughts?~

## Change checklist

- [x] I have performed a self-review of my own code
- [x] I have made corresponding changes to the documentation
- [ ] ~I have added tests that prove my fix is effective or that my feature works~
- [x] A changelog entry has been made in the appropriate crates

### CODE REVIEW

**Comment by thomaseizinger** [comment]
The inter-op tests fail because we are patching `yamux` in the entire workspace and testground doesn't seem to like that as it is now failing to compile the previous rust-libp2p version.

All local tests are green though which is exciting :blush:

**Comment by mxinden** [comment]
Cross referencing https://github.com/libp2p/rust-yamux/pull/142#issuecomment-1281274677 here for visibility.

**Comment by thomaseizinger** [comment]
@mxinden: In case you are happy with the changes, can you release this on `crates.io` please? I'll then write a "call for testing" comment on https://github.com/libp2p/rust-libp2p/issues/3202.

**Comment by thomaseizinger** [comment]
The last commit is required so that our test suite actually runs.

**Comment by rkuhn** [comment]
Since I won’t be able to do a full review of the underlying changes: is my assumption correct that the new version shall be 100% network compatible with the current version?

**Comment by thomaseizinger** [comment]
> Since I won’t be able to do a full review of the underlying changes: is my assumption correct that the new version shall be 100% network compatible with the current version?

Yes, should be 100% network compatible.

**Comment by thomaseizinger** [comment]
> @mxinden: In case you are happy with the changes, can you release this on `crates.io` please? I'll then write a "call for testing" comment on #3202.

@mxinden: Friendly ping.

**Review by mxinden — APPROVED** [review]

**Comment by mxinden** [comment]
[Tagged](https://github.com/libp2p/rust-libp2p/releases/tag/libp2p-yamux-v0.43.0-alpha) and [published](https://crates.io/crates/libp2p-yamux/0.43.0-alpha).

> I think we should probably hold off on merging

:+1:

**Comment by thomaseizinger** [comment]
@rkuhn Did you manage to run this in production without issues?

**Comment by rkuhn** [comment]
Sorry for the long delay: I currently don’t have the bandwidth to try out this change on a relevant network.

**Comment by mxinden** [comment]
@thomaseizinger can you update this pull request and propagate this change via a pull request to [kademlia-exporter](https://github.com/mxinden/kademlia-exporter/) or [rust-libp2p-server](https://github.com/mxinden/rust-libp2p-server/)? I can then deploy it to kademlia-exporter.max-inden.de/ or the rust-libp2p based IPFS bootstrap node `/dns4/ams-rust.bootstrap.libp2p.io/tcp/4001/p2p/12D3KooWEZXjE41uU4EL2gpkAQeDXYok6wghN7wwNVPF5bwkaNfS`.

**Comment by thomaseizinger** [comment]
> @thomaseizinger can you update this pull request and propagate this change via a pull request to [kademlia-exporter](https://github.com/mxinden/kademlia-exporter/) or [rust-libp2p-server](https://github.com/mxinden/rust-libp2p-server/)? I can then deploy it to kademlia-exporter.max-inden.de/ or the rust-libp2p based IPFS bootstrap node `/dns4/ams-rust.bootstrap.libp2p.io/tcp/4001/p2p/12D3KooWEZXjE41uU4EL2gpkAQeDXYok6wghN7wwNVPF5bwkaNfS`.

I guess you updated both already to the latest version so the released alpha doesn't help?

**Comment by thomaseizinger** [comment]
Actually, the released alpha depends on libp2p-core 0.38 which is compatible with libp2p 0.50.

I can send a PR nevertheless but this PR doesn't need updating :)

**Comment by thomaseizinger** [comment]
> @thomaseizinger can you update this pull request and propagate this change via a pull request to [kademlia-exporter](https://github.com/mxinden/kademlia-exporter/) or [rust-libp2p-server](https://github.com/mxinden/rust-libp2p-server/)? I can then deploy it to kademlia-exporter.max-inden.de/ or the rust-libp2p based IPFS bootstrap node `/dns4/ams-rust.bootstrap.libp2p.io/tcp/4001/p2p/12D3KooWEZXjE41uU4EL2gpkAQeDXYok6wghN7wwNVPF5bwkaNfS`.

Ah you are not depending on the released version of `libp2p` ...

**Comment by thomaseizinger** [comment]
We are kind of mixing concerns here now with this deployment strategy.

The point of an alpha release was to have _only_ yamux change and thus we could have been certain to attribute and change in behaviour to the changes in yamux. If we just deploy the latest version of `master`, we are testing a lot of changes all in one go.

Additionally, it feels a bit odd to depend on an unmerged branch in either of those two repositories.

I'd be in favor of moving one of these projects into our monorepo and deploy from here. This would allow us to also deploy individual branches (to a separate instance) and see how they perform in a live network. Any chance we can get that?

**Comment by mxinden** [comment]
> I'd be in favor of moving one of these projects into our monorepo and deploy from here. This would allow us to also deploy individual branches (to a separate instance) and see how they perform in a live network. Any chance we can get that?

I am not at all opposed to that. Less work for me.

**Comment by mxinden** [comment]
@thomaseizinger this is now running on https://kademlia-exporter.max-inden.de/d/Pfr0Fj6Mk/rust-libp2p

**Comment by thomaseizinger** [comment]
> @thomaseizinger this is now running on https://kademlia-exporter.max-inden.de/d/Pfr0Fj6Mk/rust-libp2p

Thanks! How long are we going to let it run before we consider the reported metrics to give us green light?

**Comment by mxinden** [comment]
I can not spot any anomalies from the graphs. Thus I suggest moving forward.

File: Cargo.toml
Comment (mxinden) [inline Cargo.toml]:
```suggestion
libp2p-yamux = { version = "0.43.0", path = "muxers/yamux", optional = true }
```

I suggest releasing a stable version right away.

File: muxers/yamux/CHANGELOG.md
Comment (mxinden) [inline muxers/yamux/CHANGELOG.md]: I don't think we will see any testing of the `alpha`, thus I suggest not doing the alpha release in the first place, but instead release as a new minor version right away.

**Review by mxinden — COMMENTED** [review]
@thomaseizinger can you update this pull request? Once that is done, this is ready to merge from my end.

File: muxers/yamux/CHANGELOG.md
Comment (thomaseizinger) [reply muxers/yamux/CHANGELOG.md]: We did release the alpha already :D

**Comment by thomaseizinger** [comment]
This is a breaking change too so I added it to the `0.52` milestone.

**Review by mxinden — DISMISSED** [review]
Missing version bump in `Cargo.toml` of `libp2p-yamux`, but I guess that we can do once we actually plan to merge the pull request.

**Comment by thomaseizinger** [comment]
Interestingly, we have a failure with the `js v0.41` implementation here ..

**Comment by thomaseizinger** [comment]
@maschad I am getting interop failures on this PR with js-libp2p 0.41 and 0.42. Supposedly, we are sending the wrong ACK payload back. This PR does include a big refactoring of yamux so it might still be a bug in `rust-yamux`, however it does work with all other implementations, including older versions of `rust-yamux`.

Do you know of any problems in the yamux and/or ping implementations in js-libp2p 0.41 and 0.42?

**Comment by achingbrain** [comment]
cc @wemeetagain

**Comment by MarcoPolo** [comment]
Could you try testing with the latest js-libp2p-yamux? Once I get https://github.com/libp2p/js-libp2p/pull/1615 in, it should be easy to test this.

**Comment by thomaseizinger** [comment]
> Could you try testing with the latest js-libp2p-yamux? Once I get [libp2p/js-libp2p#1615](https://github.com/libp2p/js-libp2p/pull/1615) in, it should be easy to test this.

Still fails with the latest release. See https://github.com/libp2p/test-plans/pull/178.

**Comment by thomaseizinger** [comment]
Split off the harding of API into a separate PR so we can make the version bump in a patch release later.

**Comment by thomaseizinger** [comment]
Removed this from the `v0.52` milestone in favor of https://github.com/libp2p/rust-libp2p/pull/3908. Once https://github.com/libp2p/rust-libp2p/pull/3908 is landed, we can upgrade the underlying version in a patch release, thus unblocking the next release from the issue discovered here.

File: muxers/yamux/src/lib.rs
Comment (maschad) [inline muxers/yamux/src/lib.rs]: We need to ensure that this properly flushes the buffer, see https://github.com/ChainSafe/js-libp2p-yamux/issues/48#issuecomment-1554687960

**Review by mxinden — APPROVED** [review]

File: muxers/yamux/src/lib.rs
Comment (mxinden) [inline muxers/yamux/src/lib.rs]: @thomaseizinger why rename the type parameter from `C` (I am assuming *Connection*) to `S` (I am assuming *Stream*)?

**Review by mxinden — COMMENTED** [review]
Happy to discuss after merge.

**Comment by mxinden** [comment]
Pull request is failing with nim on TCP+Noise+Yamux and WS+Noise+Yamux.

https://github.com/libp2p/rust-libp2p/actions/runs/5085921901

**Comment by thomaseizinger** [comment]
> Pull request is failing with nim on TCP+Noise+Yamux and WS+Noise+Yamux.
> 
> https://github.com/libp2p/rust-libp2p/actions/runs/5085921901

Looks like a flake? I can't see any particular error.

File: muxers/yamux/src/lib.rs
Comment (thomaseizinger) [reply muxers/yamux/src/lib.rs]: Not sure anymore :)

Don't really care!

Comment (mxinden) [reply muxers/yamux/src/lib.rs]:
```suggestion
pub struct Muxer<C> {
    connection: yamux::Connection<C>,
```

Changing to `C` to be consistent with `impl`s below.

**Comment by mxinden** [comment]
> > Pull request is failing with nim on TCP+Noise+Yamux and WS+Noise+Yamux.
> > https://github.com/libp2p/rust-libp2p/actions/runs/5085921901
> 
> Looks like a flake? I can't see any particular error.

I don't think so. It has failed in 3 consecutive runs.

**Comment by thomaseizinger** [comment]
> > > Pull request is failing with nim on TCP+Noise+Yamux and WS+Noise+Yamux.
> > > https://github.com/libp2p/rust-libp2p/actions/runs/5085921901
> > 
> > Looks like a flake? I can't see any particular error.
> 
> I don't think so. It has failed in 3 consecutive runs.

Damn. What is wrong this time 😅

**Comment by thomaseizinger** [comment]
Yey! The interop tests are finally green here!

File: interop-tests/src/lib.rs
Comment (mxinden) [inline interop-tests/src/lib.rs:42]: In case you want to merge with this change included, please add a comment.

**Review by mxinden — APPROVED** [review]
Changes (excluding dependency override) look good to me.

Thank you for debugging this across implementations!

File: interop-tests/src/lib.rs
Comment (thomaseizinger) [reply interop-tests/src/lib.rs:42]: I don't think we have a choice unless we wait until `nim-libp2p` patches its implementation.

Comment (thomaseizinger) [reply interop-tests/src/lib.rs:42]: What sort of comment would you like to see?

Comment (mxinden) [reply interop-tests/src/lib.rs:42]: A link to https://github.com/status-im/nim-libp2p/issues/924. Once that is fixed we can remove the above.

**Comment by thomaseizinger** [comment]
@mxinden can you give https://github.com/libp2p/rust-yamux/pull/153 another review?

**Comment by mxinden** [comment]
Blocked on https://github.com/libp2p/rust-yamux/issues/166.

**Comment by mxinden** [comment]
> @mxinden can you give [libp2p/rust-yamux#153](https://github.com/libp2p/rust-yamux/pull/153) another review?

Done. Approved.

**Comment by thomaseizinger** [comment]
Something is not quite right, still. For some reason, the kademlia test `two_servers_add_each_other_to_routing_table` now always fails.

It seems like some of the events now happen in a different order. Interestingly, we already seem to be receiving events for kademlia before event emitting the `ConnectionEstablished` event. I am guessing this happens because the connection tasks are actually running parallel in the background. For `Swarm::new_ephemeral` though, we don't use an executor so there is no actual background work happening.

**Comment by mxinden** [comment]
@thomaseizinger do you have time this or next week to look into the above `libp2p-kad` failures? I think realizing the many improvements in `yamux` `v0.12` in `libp2p` is worth the additional effort.

**Comment by thomaseizinger** [comment]
> @thomaseizinger do you have time this or next week to look into the above `libp2p-kad` failures? I think realizing the many improvements in `yamux` `v0.12` in `libp2p` is worth the additional effort.

I'll have a look.

**Comment by thomaseizinger** [comment]
@mxinden I adapted the tests so they are passing now. I ended up being as simple as removing two events from the match that are now (consistently) emitted earlier and thus "swallowed" by the `swarm.listen` call whilst waiting for the connection to be fully setup.

This is ready for review / merge now!

File: muxers/yamux/CHANGELOG.md
Comment (mxinden) [inline muxers/yamux/CHANGELOG.md]:
```suggestion
  When interacting with other libp2p nodes that are also running this or a newer version, the creation of inbound streams will be backpressured once the ACK backlog is hit.
```

Would be easier for me to understand.

**Review by mxinden — APPROVED** [review]
Changes look good to me. Unless the performance benchmarks (see link above) show any regressions, this is good to merge from my end.

**Comment by thomaseizinger** [comment]
Can't see a regression but also unfortunately no improvement in performance: https://observablehq.com/@libp2p-workspace/performance-dashboard?branch=perf-rust-libp2p-yamux-v0.12#branch

I am guessing we are just severely limited by the initial receive window.

**Comment by thomaseizinger** [comment]
With the above results, I am merging here!

**Comment by mxinden** [comment]
> I am guessing we are just severely limited by the initial receive window.

I am operating under the same assumption.

**Pull request merged**
